Add admin accounts

noaccOS opened the pull request:
Admins are special accounts that can create other accounts without needing confirmation, up to their admin_level and edit user accounts of a level lower than them.

In the future they will be used to insert new pets in the database.

Admins are implemented as an admin level 0..5 instead of a simple flag, to (possibly) allow more granular control over the resources.

noaccOS committed:
generate database migration

although the file is mostly empty, it was generated using
mix ecto.gen.migration add_admin_level_to_users

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
admin level migration: add field and constraint

it was decided that admin level goes from 0 (normal user)
to 5 (root user), with intermediate levels.

a user of admin_level x can only register other accounts of
admin_level <= x and edit other accounts of admin_level < x

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
initial admin work: registration

route is present but not tested yet

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
refactor: extract login information in its own function

the function loads the data in the conn's assigns

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
fix: check that creator is an admin for automatic confirmation

we don't want a normal user registering someone else and they
being marked as confirmed

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
feat: add route for user registration made by user

refactors the controller by merging create and create_admin
the two bodies are now distincted on the match of a logged in user

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
feat(fallback_controller): manage :forbidden errors

noaccOS committed:
refactor: split `register_user` in two functions

register_user doesn't perform checks, and is only used in tests and internally
try_register_user performs confirmation and admin_level checks

for this reason, the default admin_level in validate_admin_level is raised
to 5, because try_register_user will always pass the correct value and
the default is only used in register_user

noaccOS committed:
feat(accounts): add `maybe_redacted_user_by_email/2`

like fetch_user_by_email/1, but reutrns the user only if
the caller has enough permissions to view it.

otherwise returns :forbidden, as to avoid disclosing whether the
email is registered or not for another same-level or higher admin account

noaccOS committed:
feat: allow admins to request email updates for other accounts

rbino commented on priv/repo/migrations/20230508100505_add_admin_level_to_users.exs:
```diff
@@ -0,0 +1,19 @@
+defmodule PetStore.Repo.Migrations.AddAdminLevelToUsers do
+  use Ecto.Migration
+
+  def up do
```
I think you can just `s/up/change/g` removing the `down` function and Elixir will figure out how to reverse the changes. Also, we usually simply use `add` when greenfielding, the migration system should take care of checking that the migration is already applied or not.

noaccOS replied:
Unfortunately that was my starting point, but trying to `mix ecto.rollback -n 1` I get 
```
12:22:36.434 [info] drop constraint admin_level_range from table users
** (Ecto.MigrationError) cannot reverse migration command: alter table users. You will need to explicitly define up/0 and down/0 in your migration
```

I removed the `_if_exists` and `_if_not_exists` as requested in 4d1b65a80a3e7666dbb26d7db794bdf89de7796d

noaccOS replied:
Seems like using `add` instead of `add_if_not_exists`, the rollback works just fine. I merged the two functions as requested

noaccOS committed:
refactor: drop existing checks from migration

let the migration system take care of it.
should not result in a change in functionality.

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
refactor: merge `up` and `down` in `change`

if using `add` and not `add_if_not_exists`, it seems to work just fine

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

rbino commented on lib/pet_store/accounts/user.ex:
```diff
@@ -60,6 +62,15 @@ defmodule PetStore.Accounts.User do
     |> maybe_hash_password(opts)
   end
 
+  defp validate_admin_level(changeset, opts) do
+    changeset
+    |> validate_inclusion(:admin_level, 0..5, message: "invalid range")
+    |> validate_number(:admin_level,
+      less_than_or_equal_to: Keyword.get(opts, :max_admin_level, 5),
+      message: "insufficient permissions"
+    )
+  end
```
My 2 cents: you're mixing validation (i.e. "Is the `admin_level` in a valid range) with business logic rule enforcement ("Is the creator allowed to do this"). The former belongs to changesets, the latter does not.
This is also the reason why in the controller you then have to pattern match on a specific error in the changeset to transform it in `{:error, :forbidden}`. If you perform the the authz at the context level, you can just return `{:error, :forbidden}` from there.
Remember that you can always do `Ecto.Changeset.apply_changes` in the context to validate business logic rules before hitting the database

noaccOS replied:
Requested changes applied in https://github.com/noaccOS/Pet-Store/pull/4/commits/8d32459b40b133891c80259c1d89e1a553c8423c

noaccOS replied:
> Remember that you can always do `Ecto.Changeset.apply_changes` in the context to validate business logic rules before hitting the database

I think you mean `apply_action/2`?

https://hexdocs.pm/ecto/Ecto.Changeset.html#apply_changes/1
https://hexdocs.pm/ecto/Ecto.Changeset.html#apply_action/2

rbino commented on lib/pet_store_web/controllers/user_registration_controller.ex:
```diff
@@ -6,8 +6,26 @@ defmodule PetStoreWeb.UserRegistrationController do
+
+      other ->
+        other
+    end
```
See comment above

rbino commented on lib/pet_store_web/controllers/user_settings_controller.ex:
```diff
@@ -8,6 +8,31 @@ defmodule PetStoreWeb.UserSettingsController do
+    with {:ok, target} <-
+           Accounts.maybe_redacted_user_by_email(target_email, conn.assigns.current_user),
+         email_changeset = Accounts.change_user_email(target, %{email: new_email}),
+         {:ok, new_user} <- Ecto.Changeset.apply_action(email_changeset, :update) do
```
You shouldn't call `Ecto` from the Controller layer. I see there's an `apply_user_email` below, either use that or provide some other similar helper from the context.

rbino replied:
Now that I have a look at this what you probably want to do is adding an additional authz call (e.g. using Bodyguard) before `apply_user_email`. That way you don't even need `maybe_redacted_user_by_email`.

noaccOS replied:
well yeah the whole point of `maybe_redacted_user_by_email` is to provide authentication to `fetch_user_by_email`. For now I'll add a new helper function (I still don't see much of a gain in using Bodyguard), then if you insist that's the better approach I'll proceed with a refactor

noaccOS replied:
Done in https://github.com/noaccOS/Pet-Store/pull/4/commits/e0615b05e72577553d599dbaccd9bece3e5f031d

rbino commented on priv/repo/migrations/20230508100505_add_admin_level_to_users.exs:
```diff
@@ -0,0 +1,11 @@
+      add :admin_level, :integer, null: false, default: 0
+    end
+
+    create constraint(:users, :admin_level_range, check: "admin_level >= 0 AND admin_level <= 5")
```
Not totally against it, just fyi we usually keep the application constraints only at the changeset level (especially in the quick iteration phase), also because that way we have a single source of truth.

rbino commented on lib/pet_store/accounts.ex:
```diff
@@ -85,6 +97,25 @@ defmodule PetStore.Accounts do
   end
 
+  @doc """
+  Like register user, but should be preferred when registering from the web interface as
+  it checks for adequate admin_level
```
If it's "like register user", does it make sense to have them both? Technically now `register_user` is skipping some business logic checks, so I'd expect this to become the new `register_user`, given that when called with `creator == nil` it should have the same effect

noaccOS replied:
It technically is, as explained in the commit message (80086e3) `reigster_user` is there only as a shortcut for the tests.
For this reason I should've probably put it in the `accounts_fixture` file instead of the project api, will look into it in a bit to see if I can avoid it altogether

noaccOS replied:
Ok so with cc451be3305cfe2e3b7c54055f54e5e567ef92d6 I can pass all the tests in the latest branch :+1:

noaccOS committed:
refactor(users): move `max_admin_level` validation to context level

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
refactor(accounts): add `apply_user_email/2`

like `apply_user_email/3` but doesn't validate the password
used in UserSettingsController to avoid calling Ecto in the controller

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
refactor: better naming

don't call an admin_level 0 "admin"

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
refactor(accounts): re-unify `register_user`s functions

apply custom code in user_fixture to have some freedom in tests

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
fix(user): remove warning for unused `opts`

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>
